### PR TITLE
Copy expression types to avoid weight_strings and derived tables

### PR DESCRIPTION
**Pull request by wangweicugw**
## Description

The query 

```sql
select 1 from user 
union 
select 2 from user
``` 

was transformed into 

```sql
select `1`, weight_string(`1`) from (
  select 1 from `user` 
  union 
  select 2 from `user`) as dt
```

for execution, but by copying the types, it can be simplified and avoid the generation of a derived table. We can now just use the original query.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/15020

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by systay** [comment]
Hi @wangweicugw 👋 

This looks great! Thank you for the contribution.

Would you mind also adding a end-to-end test with the faulty query? It could go into the `TestUnionDistinct` method in `go/test/endtoend/vtgate/queries/union/union_test.go`

**Comment by systay** [comment]
> Would you mind also adding a end-to-end test with the faulty query? It could go into the `TestUnionDistinct` method in `go/test/endtoend/vtgate/queries/union/union_test.go`

🤦 Ignore me. I see that you already did this

File: go/vt/vtgate/planbuilder/operators/union_merging.go [hunk 1]
```diff
@@ -180,25 +180,27 @@ func createMergedUnion(
...
+		deps := ctx.SemTable.RecursiveDeps(lae.Expr)
+		rae, ok := rhsExprs[idx].(*sqlparser.AliasedExpr)
 		if !ok {
 			noDeps = true
 			continue
 		}
-		deps = deps.Merge(ctx.SemTable.RecursiveDeps(ae.Expr))
+		deps = deps.Merge(ctx.SemTable.RecursiveDeps(rae.Expr))
+		ctx.SemTable.CopySemanticInfo(rae.Expr, col)
+		ctx.SemTable.CopySemanticInfo(lae.Expr, col)
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/union_merging.go:203]: I don't think this is correct. This works when both the left and right columns have the same type, but if they have different types, we should coerce the resulting type and not just copy it.

Here is an example query:

```sql
mysql> select 1 union select 'a' union select now();
Field   1:  `1`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_0900_ai_ci (255)
Length:     76
Max_length: 19
Decimals:   0
Flags:      NOT_NULL 


+---------------------+
| 1                   |
+---------------------+
| 1                   |
| a                   |
| 2024-01-30 07:32:57 |
+---------------------+
3 rows in set (0.00 sec)
```

As you can see when you run the mysql client with the option `--column-type-info`, we have three columns with different types - int, varchar and datetime, and they are all coerced into varchar.

Fortunately, we already have this functionality today. Here is my suggestion:

```suggestion
		rt, foundR := ctx.SemTable.TypeForExpr(rae.Expr)
		lt, foundL := ctx.SemTable.TypeForExpr(lae.Expr)
		if foundR && foundL {
			types := []sqltypes.Type{rt.Type(), lt.Type()}
			t := evalengine.AggregateTypes(types)
			ctx.SemTable.ExprTypes[col] = evalengine.NewType(t, collations.Unknown)
		}
```

**Comment by systay** [comment]
We run the end-to-end tests on older binaries as well. Since this fix hasn't made it to the older versions, we need to stop the test from running if we are not on `main`. You need to extract the new test lines into it's own test method, and start this method with:

```go
utils.SkipIfBinaryIsBelowVersion(t, 19, "vtgate")
```

**Comment by wangweicugw** [comment]
> We run the end-to-end tests on older binaries as well. Since this fix hasn't made it to the older versions, we need to stop the test from running if we are not on `main`. You need to extract the new test lines into it's own test method, and start this method with:我们也对旧的二进制文件运行端到端测试。由于此修复尚未适用于旧版本，因此如果我们不在 `main` 上，我们需要停止运行测试。您需要将新的测试行提取到它自己的测试方法中，并使用以下命令启动该方法：
> 
> ```go
> utils.SkipIfBinaryIsBelowVersion(t, 19, "vtgate")
> ```
> 
> This will stop the `Query Serving (Queries) - Upgrade Downgrade Testing / Run Upgrade Downgrade Test` CI step from failing这将阻止 `Query Serving (Queries) - Upgrade Downgrade Testing / Run Upgrade Downgrade Test` CI 步骤失败

Ok👌🏻

File: go/vt/vtgate/planbuilder/operators/union_merging.go (hunk 1 above)
Comment (wangweicugw) [reply go/vt/vtgate/planbuilder/operators/union_merging.go]: Yes, I overlooked the case where the left and right columns have different types. However, in some cases, you may not be able to retrieve the ``evalengine.Type`` through ``ctx.SemTable.TypeForExpr()`` method, especially when dealing with syntax tree nodes like ``now()`` that cannot provide the evalengine.Type after processing it with ``ctx.SemTable.TypeForExpr()`` .

File: go/vt/vtgate/planbuilder/operators/union_merging.go [hunk 2]
```diff
@@ -199,8 +202,16 @@ func createMergedUnion(
...
 		deps = deps.Merge(ctx.SemTable.RecursiveDeps(rae.Expr))
-		ctx.SemTable.CopySemanticInfo(rae.Expr, col)
-		ctx.SemTable.CopySemanticInfo(lae.Expr, col)
+		rt, foundR := ctx.SemTable.TypeForExpr(rae.Expr)
+		lt, foundL := ctx.SemTable.TypeForExpr(lae.Expr)
+		if foundR && foundL {
+			types := []sqltypes.Type{rt.Type(), lt.Type()}
+			t := evalengine.AggregateTypes(types)
+			ctx.SemTable.ExprTypes[col] = evalengine.NewType(t, collations.Unknown)
+		} else {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/union_merging.go:211]: I don't think we should have the else clause here. It's not correct to use either of the two types as the common type. Like my example showed, you can have a situation where one side is an int, the other side is a datetime, but the common type is varchar.

If we don't have both types available, I think we need to use the `weight_string` function to be sure we can compare values correctly.

Comment (wangweicugw) [reply go/vt/vtgate/planbuilder/operators/union_merging.go:211]: I understand now. There shouldn't be an "else" processing logic. 

If it is a syntax tree node for the now() function, we can save its corresponding ``evalengine.Type`` in ``ExprTypes`` during the analysis phase in the analyzer.

**Review by harshit-gangal — APPROVED** [review]

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/union_merging.go [hunk 3]
```diff
@@ -180,25 +183,32 @@ func createMergedUnion(
...
 			continue
 		}
-		deps = deps.Merge(ctx.SemTable.RecursiveDeps(ae.Expr))
+		deps = deps.Merge(ctx.SemTable.RecursiveDeps(rae.Expr))
+		rt, foundR := ctx.SemTable.TypeForExpr(rae.Expr)
+		lt, foundL := ctx.SemTable.TypeForExpr(lae.Expr)
+		if foundR && foundL {
+			types := []sqltypes.Type{rt.Type(), lt.Type()}
+			t := evalengine.AggregateTypes(types)
+			ctx.SemTable.ExprTypes[col] = evalengine.NewType(t, collations.Unknown)
```
[the remark this answers is not in the document]
Comment (dbussink) [reply go/vt/vtgate/planbuilder/operators/union_merging.go:210]: Ah, https://github.com/vitessio/vitess/pull/15122 answers my question here.

**Pull request merged**